### PR TITLE
feat: add structured logging

### PR DESCRIPTION
**Pull request by sanpj2292**
## What are the changes introduced in this PR?

Adding structured logging to the service. We have used `winston` for this purpose.

## What is the related Linear task?

Resolves INT-1087

## Please explain the objectives of your changes below

Put down any required details on the broader aspect of your changes. If there are any dependent changes, **mandatorily** mention them here

### Any changes to existing capabilities/behaviour, mention the reason & what are the changes ?

N/A

### Any new dependencies introduced with this change?

N/A

### Any new generic utility introduced or modified. Please explain the changes.

N/A

### Any technical or performance related pointers to consider with the change?

N/A

<hr>

### Developer checklist

- [ ] My code follows the style guidelines of this project

- [ ] **No breaking changes are being introduced.**

- [ ] All related docs linked with the PR?

- [x] All changes manually tested?

- [ ] Any documentation changes needed with this change?

- [ ] Is the PR limited to 10 file changes?

- [x] Is the PR limited to one linear task?

- [ ] Are relevant unit and component test-cases added?

### Reviewer checklist

- [ ] Is the type of change in the PR title appropriate as per the changes?

- [ ] Verified that there are no credentials or confidential data exposed with the changes.

### CODE REVIEW

**Comment by koladilip** [comment]
We need to have discussion on this, few questions:
1. Are we planning to send a child logger with all the context to destination or source handlers? If yes how does that work?
2. I thought from root level module we need keep passing a logger to child modules functions with more context, but there is no child logger being used here. Since the context changes per request we need to create child loggers per request that may be costly (not sure if that is really concern or not) else we need to do something golang does with context, we can create child context from parent and so on, we can keep one logger at module level but context keeps changing.

**Comment by sanpj2292** [comment]
> Are we planning to send a child logger with all the context to destination or source handlers? If yes how does that work?

Most of use-case regarding logging surrounds around having a good logger that could tell us clearly about what's happening in the service. The logger requirement at destination/source level is quite less at this point. But when we consider the logger to be used at destination/source level currently this is the thought process.

> I thought from root level module we need keep passing a logger to child modules functions with more context, but there is no child logger being used here. Since the context changes per request we need to create child loggers per request that may be costly (not sure if that is really concern or not) else we need to do something golang does with context, we can create child context from parent and so on, we can keep one logger at module level but context keeps changing.

We have to look into the possibility of having a single parent logger with changing context.


As per our discussion offline, will try to implement that version & we can look at it in detail

File: src/services/destination/nativeIntegration.ts [hunk 1]
```diff
@@ -113,8 +121,12 @@ export class NativeIntegrationDestinationService implements DestinationService {
           tags.FEATURES.ROUTER,
         );
         try {
+          const metadataWithSvcCtx = {
+            ...requestMetadata,
+            serviceContext: MiscService.getLoggableData(metaTO.errorDetails),
```
Comment (koladilip) [inline src/services/destination/nativeIntegration.ts:126]: this is more of request context right, service context might be misleading?

File: src/services/destination/nativeIntegration.ts [hunk 2]
```diff
@@ -63,23 +64,30 @@ export class NativeIntegrationDestinationService implements DestinationService {
...
     const respList: ProcessorTransformationResponse[][] = await Promise.all(
       events.map(async (event) => {
+        const metaTO = this.getTags(
+          destinationType,
+          event.metadata?.destinationId,
+          event.metadata?.workspaceId,
+          tags.FEATURES.PROCESSOR,
+        );
+        metaTO.metadata = event.metadata;
+        const metadataWithSvcCtx = {
```
Comment (koladilip) [inline src/services/destination/nativeIntegration.ts:74]: I thought we can simply add to requestMetadata like following:
requestMetadata.context = {...};

File: src/services/destination/nativeIntegration.ts (hunk 1 above)
Comment (sanpj2292) [reply src/services/destination/nativeIntegration.ts:126]: The thought process behind naming was to reflect the context at service level.

File: src/services/destination/nativeIntegration.ts (hunk 2 above)
Comment (sanpj2292) [reply src/services/destination/nativeIntegration.ts:74]: We are getting an eslint error when we do as suggested hence adopted this approach

File: src/services/destination/nativeIntegration.ts (hunk 1 above)
Comment (koladilip) [reply src/services/destination/nativeIntegration.ts:126]: may be check with others and choose the one with less confusing

File: src/services/destination/nativeIntegration.ts [hunk 3]
```diff
@@ -63,23 +64,30 @@ export class NativeIntegrationDestinationService implements DestinationService {
...
+          ...requestMetadata,
+          serviceContext: MiscService.getLoggableData(metaTO.errorDetails),
```
Comment (koladilip) [inline src/services/destination/nativeIntegration.ts:76]: why only from error details? we should use 
```
destinationType,
event.metadata?.destinationId,
event.metadata?.workspaceId,
tags.FEATURES.PROCESSOR,
```

Comment (sanpj2292) [reply src/services/destination/nativeIntegration.ts:76]: We are storing the information like destinationType, destinationId etc,. at service layer level in this type. We are just re-using the same type.

**Comment by devops-github-rudderstack** [comment]
This PR is considered to be stale. It has been open for 20 days with no further activity thus it is going to be closed in 7 days. To avoid such a case please consider removing the stale label manually or add a comment to the PR.

**Pull request closed without merging**